core/chains/evm/client/errors.go

```diff
@@ -233,17 +235,62 @@ var zkEvm = ClientErrors{
 	OutOfCounters: regexp.MustCompile(`(?:: |^)not enough .* counters to continue the execution$`),
 }
 
-var clients = []ClientErrors{parity, geth, arbitrum, metis, substrate, avalanche, nethermind, harmony, besu, erigon, klaytn, celo, zkSync, zkEvm}
+var clientsLock = sync.Mutex{}
+var clients = map[string]ClientErrors{
+	"parity":     parity,
+	"geth":       geth,
+	"arbitrum":   arbitrum,
+	"metis":      metis,
+	"substrate":  substrate,
+	"avalanche":  avalanche,
+	"nethermind": nethermind,
+	"harmony":    harmony,
+	"besu":       besu,
+	"erigon":     erigon,
+	"klaytn":     klaytn,
+	"celo":       celo,
+	"zkSync":     zkSync,
+	"zkEvm":      zkEvm,
+}
+
+// SetClientErrorRegexes is called on startup to set the client errors from the config
+func SetClientErrorRegexes(errsRegex config.ClientErrors) {
+	clientsLock.Lock()
+	defer clientsLock.Unlock()
+	clients["tomlConfig"] = ClientErrors{
+		NonceTooLow:                       regexp.MustCompile(errsRegex.NonceTooLow()),
+		NonceTooHigh:                      regexp.MustCompile(errsRegex.NonceTooHigh()),
+		ReplacementTransactionUnderpriced: regexp.MustCompile(errsRegex.ReplacementTransactionUnderpriced()),
+		LimitReached:                      regexp.MustCompile(errsRegex.LimitReached()),
+		TransactionAlreadyInMempool:       regexp.MustCompile(errsRegex.TransactionAlreadyInMempool()),
+		TerminallyUnderpriced:             regexp.MustCompile(errsRegex.TerminallyUnderpriced()),
+		InsufficientEth:                   regexp.MustCompile(errsRegex.InsufficientEth()),
+		TxFeeExceedsCap:                   regexp.MustCompile(errsRegex.TxFeeExceedsCap()),
+		L2FeeTooLow:                       regexp.MustCompile(errsRegex.L2FeeTooLow()),
+		L2FeeTooHigh:                      regexp.MustCompile(errsRegex.L2FeeTooHigh()),
+		L2Full:                            regexp.MustCompile(errsRegex.L2Full()),
+		TransactionAlreadyMined:           regexp.MustCompile(errsRegex.TransactionAlreadyMined()),
+		Fatal:                             regexp.MustCompile(errsRegex.Fatal()),
+		ServiceUnavailable:                regexp.MustCompile(errsRegex.ServiceUnavailable()),
+	}
+}
 
 func (s *SendError) is(errorType int) bool {
 	if s == nil || s.err == nil {
 		return false
 	}
 	str := s.CauseStr()
+
+	clientsLock.Lock()
+	defer clientsLock.Unlock()
 	for _, client := range clients {
```

Comment:
Instead of using global state, how about passing in `clients` as an argument?
```suggestion
	for _, client := range clients {
```
It could be required, optional to override the default, or supplemental to consider in addition to (or before?) the defaults.

Another option to consider would be to refactor this to be a method on `ClientErrors` instead:
```go
func (e *ClientErrors) errIs(err error, errorType int) bool
```
The the caller doesn't have to understand a complex argument - then can just use the `ClientErrors` they want however they need to.

Reply:
Please reconsider this or explain the advantage of carrying around the `ClientErrors` as field. It seems like an inversion of the natural relationship and entangles separate concerns.

Reply:
Yeah I see what you mean, I refactored it as you suggested.